zerver/lib/push_notifications.py

```diff
@@ -603,6 +603,13 @@ def get_apns_alert_subtitle(message: Message) -> str:
     # For group PMs, or regular messages to a stream, just use a colon to indicate this is the sender.
     return message.sender.full_name + ":"
 
+def get_apns_badge_counts(user_profile: UserProfile) -> int:
+    push_notification_ids = list(UserMessage.objects.filter(
+        user_profile=user_profile,
+        flags=UserMessage.flags.active_mobile_push_notification).order_by(
```

Comment:
This `flags=` condition doesn't do the right thing at all.

Take a look at the one place we currently use this flag, which is in `do_clear_mobile_push_notifications_for_ids`, the function that causes `remove` notification messages to get sent. (This query kind of looks like you already based it on that code, in which case take another look.)

```python
    filtered_message_ids = list(UserMessage.objects.filter(
        message_id__in=message_ids,
        user_profile=user_profile,
    ).extra(
        where=[UserMessage.where_active_push_notification()],
    ).values_list('message_id', flat=True))
```

Notice the `where_active_push_notification`.

There is a reason it is that way and not this other way. If you take a look at the implementation of that method, and read the comments, they explain why.

This is kind of a tricky area and it's perfectly reasonable not to understand it on first look, even after reading those. #backend would be a good place to ask questions about it. What's not a good idea is to change it from one thing to another without understanding and without explaining why.

Reply:
I wonder why the tests passed despite this. That could be an indicator that there's something more we should be verifying in them, because we'd definitely want to catch using an incorrect query with them.

Reply:
@mateuszmandera +1. The function was giving the correct value for all the test cases.

Reply:
Like I said in the top-level review comment, the unit tests in this PR certainly are inadequate.

And that is a problem with this PR, not with the existing code. (Even if there are more tests it'd be good for the existing code to have, as there probably are.) When you add new functionality, you should expect to add new tests for it. Saying "but it passed all the existing tests" is a bad excuse.

@hashirsarwar The thing Mateusz is suggesting is a thing for *you* to do. :slightly_smiling_face: Please try doing it -- please take a look at the underlying code and try to work out what situation would test that difference, so that you can write a test that exercises that. If you need help understanding it (which, again, would be perfectly reasonable), ask questions in #backend and I or Tim or Steve will be glad to answer.

Reply:
Yeah, I know it requires more unit tests. I'll definitely work on it. I didn't say it as an excuse, I just agreed with what @mateuszmandera suggested to me. :slightly_smiling_face: 
Thanks once again for the feedback @gnprice @mateuszmandera!

Reply:
> What's not a good idea is to change it from one thing to another without understanding and without explaining why.

@gnprice I know that's not a good idea and I didn't do that here. If you look at the [issue](https://github.com/zulip/zulip/issues/10271) itself, Tim already suggested from where to take the code from i.e `get_mobile_push_notification_ids(user_profile)` which is a test function. And it looks like the function uses the same `flag=` condition as I used previously.

Reply:
> where to take the code from i.e `get_mobile_push_notification_ids(user_profile)` which is a test function and I used the same function

Interesting, thanks.

In general it's always a good idea to look for where something is used in non-test code, because test code can often do things that wouldn't be a good idea to do in the actual app. That's why when I searched for uses of this flag while reviewing this PR, I didn't even particularly notice that test function, and described `do_clear_mobile_push_notifications_for_ids` as "the one place we currently use this flag".

Here, it would probably be good to add a small commit editing that test function to use `where_active_push_notification` too. That way it's consistent with the app code and avoids potentially setting a bad example for the next person.

Reply:
That makes sense. I'll add a commit for that shortly. Thanks.